ec/src/scalar_mul/glv.rs

```diff
@@ -1,62 +1,163 @@
-    const COEFF_C2: Self::BaseField;
-
-    // Constants for scalar decomposition.
-    // This is a 2x2 matrix, which is practically the LLL-reduced bases.
-
-    /// The first element of the matrix for scalar decomposition.
-    const COEFF_N11: Self::ScalarField;
-    /// The second element of the matrix for scalar decomposition.
-    const COEFF_N12: Self::ScalarField;
-    /// The third element of the matrix for scalar decomposition.
-    const COEFF_N21: Self::ScalarField;
-    /// The forth element of the matrix for the scalar decomposition.
-    const COEFF_N22: Self::ScalarField;
-
-    /// Maps a point G to phi(G):= lambda G where psi is the endomorphism.
-    // On an affine curve, the function takes the following steps:
-    //  f(y) = a_1 * (y + a_2) * (y + a_3)
-    //  g(y) = b_1 * (y + b_2) * (y + b_3)
-    //  h(y) = (y + c_1) * (y + c_2)
-    //  return (x',y') where
-    //  x' = x * f(y) / y
-    //  y' = g(y) / h(y)
-    fn endomorphism(
-        base: &<Self::Curve as CurveGroup>::Affine,
-    ) -> <Self::Curve as CurveGroup>::Affine;
+pub trait GLVConfig: Send + Sync + 'static + SWCurveConfig {
+    /// Constants that are used to calculate `phi(G) := lambda*G`.
+
+    /// The coefficients of the endomorphism
+    const ENDO_COEFFS: &'static [Self::BaseField];
+
+    /// The eigenvalue corresponding to the endomorphism.
+    const LAMBDA: Self::ScalarField;
+
+    /// A 4-element vector representing a 2x2 matrix of coefficients the for scalar decomposition, s.t. k-th entry in the vector is at col i, row j in the matrix, with ij = BE binary decomposition of k.
+    /// The entries are the LLL-reduced bases.
+    /// The determinant of this matrix must equal `ScalarField::characteristic()`.
+    const SCALAR_DECOMP_COEFFS: [(bool, Self::ScalarField); 4];
 
     /// Decomposes a scalar s into k1, k2, s.t. s = k1 + lambda k2,
-    fn scalar_decomposition(k: &Self::ScalarField) -> (Self::ScalarField, Self::ScalarField);
+    fn scalar_decomposition(
+        k: Self::ScalarField,
+    ) -> (
+        <Self as CurveConfig>::ScalarField,
+        bool,
+        <Self as CurveConfig>::ScalarField,
+        bool,
+    ) {
+        let scalar: BigInt = k.into_bigint().into().into();
+
+        let coeff_bigints: [BigInt; 4] = Self::SCALAR_DECOMP_COEFFS
+            .iter()
+            .map(|x| {
+                BigInt::from_biguint(
+                    x.0.then(|| Sign::Plus).unwrap_or_else(|| Sign::Minus),
+                    x.1.into(),
+                )
+            })
+            .collect::<Vec<_>>()
+            .try_into()
+            .unwrap();
+
+        let [n11, n12, n21, n22] = coeff_bigints;
+
+        let r = BigInt::from(Self::ScalarField::MODULUS.into());
+
+        // beta = vector([k,0]) * self.curve.N_inv
+        // The inverse of N is 1/r * Matrix([[n22, -n12], [-n21, n11]]).
+        // so β = (k*n22, -k*n12)/r
+
+        let beta_1 = &scalar * &n22 / &r;
+        let beta_2 = &scalar * &n12 / &r;
```

Comment:
Let's leave a TODO to add multiplication and division methods to `BigInteger` so that we can avoid this allocating code.

Reply:
https://github.com/arkworks-rs/algebra/issues/645